places/src/storage.rs

```diff
@@ -0,0 +1,302 @@
+            let guid = super::sync::util::random_guid().expect("according to logins-sql, this is fine :)");
+            let sql = "INSERT INTO moz_places (guid, url, url_hash)
+                       VALUES (:guid, :url, hash(:url))";
+            db.execute_named_cached(sql, &[
+                (":guid", &guid),
+                (":url", &url.clone().into_string()),
+            ])?;
+            Ok(PageInfo {
+                page_id: PageId::Guid(SyncGuid(guid)),
+                row_id: RowId(db.db.last_insert_rowid()),
+                url: url.clone(),
+                title: "".into(),
+                hidden: true, // will be set to false as soon as a non-hidden visit appears.
+                typed: 0,
+                frecency: -1,
+                visit_count_local: 0,
+                visit_count_remote: 0,
+                last_visit_date_local: Timestamp(0),
+                last_visit_date_remote: Timestamp(0),
+            })
+        }
+    }
+}
+
+// Add a single visit - you must know the page rowid. Does not update the
+// page info - if you are calling this, you will also need to update the
+// parent page with the new visit count, frecency, etc.
+fn add_visit(db: &PlacesDb,
+                 page_id: &RowId,
+                 from_visit: &Option<RowId>,
+                 visit_date: &Timestamp,
+                 visit_type: &VisitTransition,
+                 is_local: &bool) -> Result<RowId> {
+    let sql =
+        "INSERT INTO moz_historyvisits
+            (from_visit, place_id, visit_date, visit_type, is_local)
+        VALUES (:from_visit, :page_id, :visit_date, :visit_type, :is_local)";
+    db.execute_named_cached(sql, &[
+        (":from_visit", from_visit),
+        (":page_id", page_id),
+        (":visit_date", visit_date),
+        (":visit_type", visit_type),
+        (":is_local", is_local),
+    ])?;
+    let rid = db.db.last_insert_rowid();
+    Ok(RowId(rid))
+}
+
+// Currently not used - we update the frecency as we update the page info.
+pub fn update_frecency(db: &PlacesDb, id: RowId, redirect: Option<bool>) -> Result<()> {
+    let score = frecency::calculate_frecency(&db,
+        &frecency::DEFAULT_FRECENCY_SETTINGS,
+        id.0, // TODO: calculate_frecency should take a RowId here.
+        redirect)?;
+
+    db.execute_named("
+        UPDATE moz_places
+        SET frecency = :frecency
+        WHERE id = :page_id",
+        &[(":frecency", &score), (":page_id", &id.0)])?;
+
+    Ok(())
+}
+
+// Mini experiment with an "Origin" object that knows how to rev_host() itself,
+// that I don't want to throw away yet :) I'm really not sure exactly how
+// moz_origins fits in TBH :/
+#[cfg(test)]
+mod tests {
+    use unicode_segmentation::UnicodeSegmentation;
+
+    struct Origin {
+        prefix: String,
+        host: String,
+        frecency: i64,
+    }
+    impl Origin {
+        pub fn rev_host(&self) -> String {
+            self.host.graphemes(true).rev().flat_map(|g| g.chars()).collect()
```

Comment:
I suspect we should be operating on the [ascii serialization](https://docs.rs/url/1.7.1/url/enum.Origin.html#method.ascii_serialization) of the origin here.

If we're not, we should check that we really should be operating on graphemes, however. Note that the results of a a by-grapheme reversal depend on the unicode version. For many cases that's what you want, but for this I'm unsure.